tests/test_xopen.py

```diff
@@ -89,40 +108,74 @@ def fname(request):
     return request.param
 
 
-@pytest.fixture
-def lacking_pigz_permissions(tmp_path):
+@pytest.fixture(params=ALL_READERS_WITH_EXTENSION)
+def reader(request):
+    return request.param
+
+
+@pytest.fixture(params=THREADED_READERS)
+def threaded_reader(request):
+    return request.param
+
+
+@pytest.fixture(params=ALL_WRITERS_WITH_EXTENSION)
+def writer(request):
+    return request.param
+
+
+@contextmanager
+def disable_binary(tmp_path, binary_name):
     """
-    Set PATH to a directory that contains a pigz binary with permissions set to 000.
-    If no suitable pigz binary could be found, PATH is set to an empty directory
+    Find the location of the binary by its name, then set PATH to a directory that contains
+    the binary with permissions set to 000. If no suitable binary could be found,
+    PATH is set to an empty directory
     """
-    pigz_path = shutil.which("pigz")
-    if pigz_path:
-        shutil.copy(pigz_path, str(tmp_path))
-        os.chmod(str(tmp_path / "pigz"), 0)
+    try:
+        binary_path = shutil.which(binary_name)
+        if binary_path:
+            shutil.copy(binary_path, str(tmp_path))
+            os.chmod(str(tmp_path / binary_name), 0)
+        path = os.environ["PATH"]
+        os.environ["PATH"] = str(tmp_path)
+        yield
+    finally:
+        os.environ["PATH"] = path
 
-    path = os.environ["PATH"]
-    os.environ["PATH"] = str(tmp_path)
-    yield
-    os.environ["PATH"] = path
+
+@pytest.fixture
+def lacking_pigz_permissions(tmp_path):
+    with disable_binary(tmp_path, "pigz"):
+        yield
 
 
 @pytest.fixture
-def large_gzip(tmpdir):
-    path = str(tmpdir.join("large.gz"))
-    random_text = ''.join(random.choice('ABCDEFGHIJKLMNOPQRSTUVWXYZ\n') for _ in range(1024))
-    # Make the text a lot bigger in order to ensure that it is larger than the
-    # pipe buffer size.
-    random_text *= 1024
-    with xopen(path, 'w') as f:
-        f.write(random_text)
-    return path
+def lacking_pbzip2_permissions(tmp_path):
+    with disable_binary(tmp_path, "pbzip2"):
+        yield
 
 
 @pytest.fixture
-def truncated_gzip(large_gzip):
-    with open(large_gzip, 'a') as f:
-        f.truncate(os.stat(large_gzip).st_size - 10)
-    return large_gzip
+def create_large_file(tmpdir):
+    def _create_large_file(extension):
+        path = str(tmpdir.join(f"large{extension}"))
+        random_text = ''.join(random.choice('ABCDEFGHIJKLMNOPQRSTUVWXYZ\n') for _ in range(1024))
+        # Make the text a lot bigger in order to ensure that it is larger than the
+        # pipe buffer size.
+        random_text *= 1024
```

Comment:
@marcelm If I increase this value to 2048, `test_reader_close` hangs. I think this is a different issue to the incorrect placement of the `message = self._stderr.read().strip()`. This time, the test also hangs on linux. I think pbzip2 does not react to SIGTERM because the process is still running and it causes `self.process.wait()` to hang (I confirmed it by adding a timeout of 10 secongs with caused TimeoutExpired

Reply:
It does react to SIGKILL though.. which would require `self.process.kill()`